include/nbl/builtin/hlsl/cpp_compat/impl/intrinsics_impl.hlsl

```diff
@@ -3,428 +3,573 @@
-struct lerp_helper<T, bool>
+#ifndef __HLSL_VERSION
+template<typename EnumT>
+requires std::is_enum_v<EnumT>
+struct bitCount_helper<EnumT>
 {
-	static inline T lerp(NBL_CONST_REF_ARG(T) x, NBL_CONST_REF_ARG(T) y, NBL_CONST_REF_ARG(bool) a)
+	using return_t = int32_t;
+	using underlying_t = std::underlying_type_t<EnumT>;
+	static return_t __call(NBL_CONST_REF_ARG(EnumT) val)
 	{
-		if (a)
-			return y;
-		else
-			return x;
+		return bitCount_helper<const underlying_t>::__call(reinterpret_cast<const underlying_t&>(val));
 	}
 };
+#endif
 
-template<typename T, int N>
-struct lerp_helper<vector<T, N>, vector<bool, N> >
+template<typename BooleanVector>
+NBL_PARTIAL_REQ_TOP(is_vector_v<BooleanVector>&& is_same_v<typename vector_traits<BooleanVector>::scalar_type, bool>)
+struct all_helper<BooleanVector NBL_PARTIAL_REQ_BOT(is_vector_v<BooleanVector>&& is_same_v<typename vector_traits<BooleanVector>::scalar_type, bool>) >
 {
-	using output_vec_t = vector<T, N>;
-
-	static inline output_vec_t lerp(NBL_CONST_REF_ARG(output_vec_t) x, NBL_CONST_REF_ARG(output_vec_t) y, NBL_CONST_REF_ARG(vector<bool, N>) a)
+	static bool __call(NBL_CONST_REF_ARG(BooleanVector) x)
 	{
-		output_vec_t retval;
-		for (uint32_t i = 0; i < vector_traits<output_vec_t>::Dimension; i++)
-			retval[i] = a[i] ? y[i] : x[i];
-		return retval;
-	}
-};
+		using traits = hlsl::vector_traits<BooleanVector>;
+		array_get<BooleanVector, typename traits::scalar_type> getter;
+		array_set<BooleanVector, typename traits::scalar_type> setter;
 
-template<typename Matrix>
-struct transpose_helper;
-
-template<typename T, int N, int M>
-struct transpose_helper<matrix<T, N, M> >
-{
-	using transposed_t = typename matrix_traits<matrix<T, N, M> >::transposed_type;
+		bool output = true;
+		for (uint32_t i = 0; i < traits::Dimension; ++i)
+			output = output && getter(x, i);
 
-	static transposed_t transpose(NBL_CONST_REF_ARG(matrix<T, N, M>) m)
-	{
-#ifdef __HLSL_VERSION
-		return spirv::transpose(m);
-#else
-		return reinterpret_cast<transposed_t&>(glm::transpose(reinterpret_cast<typename matrix<T, N, M>::Base const&>(m)));
-#endif
+		return output;
 	}
 };
 
-template<typename LhsT, typename RhsT>
-struct mul_helper
+template<typename BooleanVector>
+NBL_PARTIAL_REQ_TOP(is_vector_v<BooleanVector>&& is_same_v<typename vector_traits<BooleanVector>::scalar_type, bool>)
+struct any_helper<BooleanVector NBL_PARTIAL_REQ_BOT(is_vector_v<BooleanVector>&& is_same_v<typename vector_traits<BooleanVector>::scalar_type, bool>) >
 {
-	static inline RhsT multiply(LhsT lhs, RhsT rhs)
+	static bool __call(NBL_CONST_REF_ARG(BooleanVector) x)
 	{
-		return mul(lhs, rhs);
+		using traits = hlsl::vector_traits<BooleanVector>;
+		array_get<BooleanVector, typename traits::scalar_type> getter;
+		array_set<BooleanVector, typename traits::scalar_type> setter;
+
+		bool output = false;
+		for (uint32_t i = 0; i < traits::Dimension; ++i)
+			output = output || getter(x, i);
+
+		return output;
 	}
 };
```

Comment:
why not Vectorial of bool ?

Reply:
btw where are the default specializations for vector in `_hLSL_VERSION`  for any and all ?